### PR TITLE
Added a check for passed null from htsjdk for .fasta files missing dictionary files

### LINKED ISSUE #2020: A fasta with a missing dict/fai can trigger a NullPointerException on Spark
Reported by Intel. Should be a `UserException` instead.

```
./gatk-launch BwaSpark -I hdfs://sn1:8020/user/$USER/gatk/CEUTrio.HiSeq.WEx.b37.NA12892.readnamesort.bam -R hdfs://sn1:8020/user/$USER/gatk/human_g1k_v37.fasta -O hdfs://sn1:8020/user/$USER/gatk/CEUTrio.HiSeq.WEx.b37.NA12892.readnamesort.bam -- --sparkRunner SPARK --sparkMaster spark://sn1:7077 --driver-memory 8G --num-executors 4 --executor-cores 9 --executor-memory 27g
 ```
 
```
java.lang.NullPointerException
        at org.broadinstitute.hellbender.utils.SequenceDictionaryUtils.getContigNames(SequenceDictionaryUtils.java:464)
        at org.broadinstitute.hellbender.utils.SequenceDictionaryUtils.getCommonContigsByName(SequenceDictionaryUtils.java:458)
        at org.broadinstitute.hellbender.utils.SequenceDictionaryUtils.compareDictionaries(SequenceDictionaryUtils.java:234)
        at org.broadinstitute.hellbender.utils.SequenceDictionaryUtils.validateDictionaries(SequenceDictionaryUtils.java:150)
        at org.broadinstitute.hellbender.utils.SequenceDictionaryUtils.validateDictionaries(SequenceDictionaryUtils.java:98)
        at org.broadinstitute.hellbender.engine.spark.GATKSparkTool.validateToolInputs(GATKSparkTool.java:402)
        at org.broadinstitute.hellbender.engine.spark.GATKSparkTool.runPipeline(GATKSparkTool.java:312)
        at org.broadinstitute.hellbender.engine.spark.SparkCommandLineProgram.doWork(SparkCommandLineProgram.java:38)
        at org.broadinstitute.hellbender.cmdline.CommandLineProgram.runTool(CommandLineProgram.java:108)
        at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMainPostParseArgs(CommandLineProgram.java:166)
        at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:185)
        at org.broadinstitute.hellbender.Main.instanceMain(Main.java:76)
        at org.broadinstitute.hellbender.Main.main(Main.java:92)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:731)
        at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:181)
        at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:206)
        at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:121)
        at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
 ```

### PR DESCRIPTION
**Pull request by jamesemery**
resolves #2020

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/engine/spark/GATKSparkTool.java [hunk 1]
```diff
@@ -352,6 +352,9 @@ private void initializeReference() {
         if ( referenceURL != null ) {
             referenceSource = new ReferenceMultiSource(gcsOptions, referenceURL, getReferenceWindowFunction());
             referenceDictionary = referenceSource.getReferenceSequenceDictionary(readsHeader != null ? readsHeader.getSequenceDictionary() : null);
+            if (referenceDictionary == null) {
+                throw new UserException.MissingReferenceDictFile(referenceURL);
+            }
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/GATKSparkTool.java]: Add an integration test showing that this exception gets thrown for a Spark tool given a fasta input with a missing `.dict` file.

**Comment by droazen** [comment]
Back to @jamesemery -- needs an integration test.

**Comment by jamesemery** [comment]
@droazen Added a test to BQSR spark

File: src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/BQSRPipelineSparkIntegrationTest.java [hunk 2]
```diff
@@ -132,4 +132,19 @@ public void testBlowUpOnBroadcastIncompatibleReference() throws IOException {
                 UserException.Require2BitReferenceForBroadcast.class);
         spec.executeTest("testBQSR-" + params.args, this);
     }
+
+    @Test(expectedExceptions = UserException.MissingReferenceDictFile.class)
+    public void testMissingReferenceDictFileCatch() throws IOException {
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/BQSRPipelineSparkIntegrationTest.java]: It's best not to bury tests for engine features in tests for tools like this (BQSR, in this case). Recommend moving to a new class `GATKSparkToolIntegrationTest` in the `engine.spark` package.

File: src/test/resources/org/broadinstitute/hellbender/tools/BQSR/validNoDict.fasta [hunk 3]
```diff
@@ -0,0 +1,40 @@
+>chr1
```
Comment (droazen) [inline src/test/resources/org/broadinstitute/hellbender/tools/BQSR/validNoDict.fasta]: Move test data to `engine/spark` as well

**Comment by droazen** [comment]
Back to @jamesemery for a simple change

**Comment by jamesemery** [comment]
@droazen made the changes you suggested

File: src/test/java/org/broadinstitute/hellbender/engine/spark/GATKSparkToolIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class GATKSparkToolIntegrationTest extends CommandLineProgramTest {
+
+    @CommandLineProgramProperties(
+            summary = "Dummy empty command line that requires a reference .",
+            oneLineSummary = "empty class",
+            programGroup = SparkProgramGroup.class
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/GATKSparkToolIntegrationTest.java]: `TestProgramGroup.class`

**Comment by droazen** [comment]
:+1: one comment, then squash/rebase/merge

**Pull request merged**